pkg/route/api/validation/validation.go

```diff
@@ -102,9 +108,121 @@ func validateTLS(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
 		if len(tls.DestinationCACertificate) > 0 {
 			result = append(result, fielderrors.NewFieldInvalid("destinationCACertificate", tls.DestinationCACertificate, "edge termination does not support destination certificates"))
 		}
+
+		result = append(result, validateTLSCertificates(tls)...)
+
 	default:
 		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or emtpy (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
 		result = append(result, fielderrors.NewFieldInvalid("termination", tls.Termination, msg))
 	}
 	return result
 }
+
+// validateTLSCertificates checks that the certificates passed are able to be parsed.  This includes ensuring that
+// the cert/key pair are matching and, if a ca cert is provided it will check the first ca cert against the leaf
+// to ensure that the leaf was signed by the ca
+func validateTLSCertificates(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
+	if tls == nil {
+		return nil
+	}
+	//edge and reencrypt (the use cases that call this require all 3 (at least) so we can ignore anything
+	//that doesn't have all 3 set.  It will already fail validation elsewhere
+	if len(tls.Certificate) == 0 || len(tls.Key) == 0 || len(tls.CACertificate) == 0 {
+		return nil
+	}
+
+	result := fielderrors.ValidationErrorList{}
+
+	//check cert/key pair
+	checkSignature := true
+	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
```

Comment:
Well, the router should be robust to things it can't handle (each router).  I'm not totally opposed to validation, but I have a healthy respect for the differing formats other crypto tools *require*, so we need to think about the implications.  Can you look for evidence this could be a problem as is?

----- Original Message -----
> > +// to ensure that the leaf was signed by the ca
> > +func validateTLSCertificates(tls *routeapi.TLSConfig)
> > fielderrors.ValidationErrorList {
> > +	if tls == nil {
> > +		return nil
> > +	}
> > +	//edge and reencrypt (the use cases that call this require all 3 (at
> > least) so we can ignore anything
> > +	//that doesn't have all 3 set.  It will already fail validation elsewhere
> > +	if len(tls.Certificate) == 0 || len(tls.Key) == 0 ||
> > len(tls.CACertificate) == 0 {
> > +		return nil
> > +	}
> > +
> > +	result := fielderrors.ValidationErrorList{}
> > +
> > +	//check cert/key pair
> > +	checkSignature := true
> > +	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
> 
> Are you thinking that we shouldn't validate the submitted certs at all?  In
> that case, the invalid cert will need to be deleted before the router is
> back online.
> 
> We could try to design a rollback mechanism in the router plugin that would
> revert if the restart command returns with a failure as an alternative to up
> front validation.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1824/files#r28774008